crates/ty_python_semantic/src/types/signatures.rs

```diff
@@ -533,83 +584,45 @@ impl<'db> Signature<'db> {
-                self.current_other = self.iter_other.next();
-                self.current_other
-            }
-
-            /// Peek at the next parameter in the `other` parameter iterator without consuming it.
-            fn peek_other(&mut self) -> Option<&'a Parameter<'db>> {
-                self.iter_other.clone().next()
-            }
-
-            /// Consumes the `ParametersZip` and returns a two-element tuple containing the
-            /// remaining parameters in the `self` and `other` iterators respectively.
-            ///
-            /// The returned iterators starts with the current parameter, if any, followed by the
-            /// remaining parameters in the respective iterators.
-            fn into_remaining(
-                self,
-            ) -> (
-                impl Iterator<Item = &'a Parameter<'db>>,
-                impl Iterator<Item = &'a Parameter<'db>>,
-            ) {
-                (
-                    self.current_self.into_iter().chain(self.iter_self),
-                    self.current_other.into_iter().chain(self.iter_other),
-                )
-            }
-        }
+        self.compare_with(db, other, &mut SignatureCheckMode::Relation(relation))
+            .unwrap_or(false)
+    }
 
-        let check_types = |type1: Option<Type<'db>>, type2: Option<Type<'db>>| {
+    /// This function is the implementation for both `has_relation_to` and
+    /// `infer_generic_specialization`. The behavior is controlled by the `mode` parameter.
+    fn compare_with<'a>(
+        &self,
+        db: &'db dyn Db,
+        other: &Signature<'db>,
+        mode: &mut SignatureCheckMode<'a, 'db>,
+    ) -> Result<bool, SpecializationError<'db>> {
+        let check_types = |type1: Option<Type<'db>>, type2: Option<Type<'db>>, relation| {
             type1.unwrap_or(Type::unknown()).has_relation_to(
                 db,
                 type2.unwrap_or(Type::unknown()),
                 relation,
             )
         };
 
-        // Return types are covariant.
-        if !check_types(self.return_ty, other.return_ty) {
-            return false;
+        match mode {
+            SignatureCheckMode::Relation(relation) => {
+                // Return types are covariant.
+                if !check_types(self.return_ty, other.return_ty, *relation) {
+                    return Ok(false);
+                }
+            }
+            SignatureCheckMode::InferSpecialization(builder) => {
+                // Return types are covariant. For inference, `self` is formal, `other` is actual.
+                if let (Some(self_type), Some(other_type)) = (self.return_ty, other.return_ty) {
+                    builder.infer(self_type, other_type)?;
+                }
+            }
         }
 
         // A gradual parameter list is a supertype of the "bottom" parameter list (*args: object,
         // **kwargs: object).
-        if other.parameters.is_gradual()
+        if mode.is_relation()
```

Comment:
(This check, for instance, is morally equivalent to a `match mode` and so should ideally move into `SignatureMatchMode` as well)